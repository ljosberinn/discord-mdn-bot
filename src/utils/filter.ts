/**
 * same as Array#filter but works on all iterables, returns an iterable
 *
 * @export
 * @template T input iterable
 * @template U output iterable
 * @param {(item: T, index: number, iter: Iterable<T>) => any} fn
 * @param {Iterable<T>} iter
 */
export function* filterʹ<T>(
  fn: (item: T, index: number, iter: Iterable<T>) => T,
  iter: Iterable<T>
) {
  let i = 0;
  for (const item of iter) {
    if (fn(item, i++, iter)) yield item;
  }
}

export function filter<T, U>(
  fn: (item: T, index: number, iter: Iterable<T>) => U
) {
  return (iter: Iterable<T>) => filterʹ(fn, iter);
}
