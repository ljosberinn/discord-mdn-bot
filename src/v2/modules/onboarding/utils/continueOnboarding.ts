import type { GuildMember, Guild, TextChannel } from 'discord.js';

import { ONBOARDING_CHANNEL } from '../../../env';
import type { UserStateType } from '../db/user_state';
import { handleIntroduction } from '../steps/handleIntroduction';
import { handleOnboarded } from '../steps/handleOnboarded';
import { handleRoleSelection } from '../steps/handleRoleSelection';
import { handleStart } from '../steps/handleStart';

export async function continueOnboarding(
  guild: Guild,
  member: GuildMember,
  oldState: UserStateType,
  fromStart = false
): Promise<void> {
  const channel = guild.channels.resolve(ONBOARDING_CHANNEL) as TextChannel;

  switch (oldState.state) {
    case 'START':
      handleStart(guild, member, oldState, fromStart);
      return;
    case 'INTRODUCTION':
      handleIntroduction(guild, member, oldState, fromStart);
      return;
    case 'ROLE_SELECTION':
      handleRoleSelection(guild, member, oldState, fromStart);
      return;
    case 'ONBOARDED':
      handleOnboarded(guild, member, oldState, fromStart);
      return;
    default:
      console.error("Shouldn't have gotten here D:");
  }
}
