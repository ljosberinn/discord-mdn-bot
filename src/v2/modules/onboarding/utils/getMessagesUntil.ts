import type { Message, TextChannel } from 'discord.js';
import { filter } from 'domyno';

export async function* getMessagesUntil(
  channel: TextChannel,
  date: Date,
  limit = 100
): AsyncIterableIterator<Message> {
  let before;
  const timestamp = date.getTime();

  const msgAfter = filter<Message>(
    (msg: Message) => msg.createdTimestamp > timestamp
  );

  while (true) {
    // this is fine this is a async generator function
    // eslint-disable-next-line no-await-in-loop
    const messages = await channel.messages.fetch({ limit, before });
    if (messages.size === 0) {
      return;
    }
    yield* msgAfter(messages.values());
    before = messages.last().id;
  }
}
