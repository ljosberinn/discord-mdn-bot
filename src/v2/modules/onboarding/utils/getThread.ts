import type { Guild, TextChannel, ThreadChannel } from 'discord.js';

import { ONBOARDING_CHANNEL } from '../../../env';

export async function getThread(
  guild: Guild,
  id: string
): Promise<ThreadChannel> {
  const channel = (await guild.channels.fetch(
    ONBOARDING_CHANNEL
  )) as TextChannel;

  return channel.threads.fetch(id);
}
