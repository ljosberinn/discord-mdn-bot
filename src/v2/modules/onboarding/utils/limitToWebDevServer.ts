import { SERVER_ID } from '../../../env';

export function limitToWebDevServer<
  HasGuildId extends { guild: { id: string } },
  Output,
  VArgs extends ReadonlyArray<unknown>
>(
  fn: (g: HasGuildId, ...vargs: VArgs) => Output
): (a: HasGuildId, ...vargs: VArgs) => Output {
  return (x: HasGuildId, ...args: VArgs) => {
    if (x.guild.id !== SERVER_ID) {
      return;
    }

    return fn(x, ...args);
  };
}
