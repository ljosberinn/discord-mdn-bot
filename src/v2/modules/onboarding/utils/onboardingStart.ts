import { get, upsert } from '../../../cache';
import type { GenericCacheType } from '../../../cache/model';
import { SERVER_ID } from '../../../env';

type OnboardingStartCache = {
  meta: {
    onboardingStart: number;
  };
} & GenericCacheType;

const _getOnboardingCache = (): Promise<OnboardingStartCache> =>
  get({
    guild: SERVER_ID,
    type: 'ONBOARDING_START',
    user: '',
  }) as unknown as Promise<OnboardingStartCache>;

export const getOnboardingStart = async (): Promise<number | undefined> => {
  const cache = await _getOnboardingCache();

  return cache?.meta.onboardingStart;
};

export const setOnboardingStart = (): Promise<OnboardingStartCache> =>
  upsert({
    expiresAt: Number.MAX_SAFE_INTEGER,
    guild: SERVER_ID,
    type: 'ONBOARDING_START',
    user: '',
    meta: {
      onboardingStart: Date.now(),
    },
  }) as unknown as Promise<OnboardingStartCache>;
