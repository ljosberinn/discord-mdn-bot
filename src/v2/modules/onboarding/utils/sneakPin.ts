import type {Message} from 'discord.js'

export const sneakPin = async (msg: Message): Promise<void> => {
  const awaitedPinned = msg.channel.awaitMessages({
    filter: x => x.type === 'CHANNEL_PINNED_MESSAGE',
    max: 1
  })

  await msg.pin()

  const pinned = await awaitedPinned
  await pinned.first().delete()
}
