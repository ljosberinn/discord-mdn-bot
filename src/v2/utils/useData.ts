import { addBreadcrumb } from '@sentry/node';
import type {
  HeadersInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';
import fetch from 'node-fetch';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env.js';
import { Cache } from './Cache.js';

const apiCache = new Cache({
  checkperiod: Number.parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  maxKeys: Number.parseInt(API_CACHE_ENTRIES_LIMIT, 10),
  stdTTL: Number.parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper):FetchWithFormat<TParsedResponse> => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get(casedCacheKey);

  if (cachedResponse) {
    return (async () => cachedResponse) as FetchWithFormat<TParsedResponse>;
  }

  return async (url, fetchOptions) => {
    addBreadcrumb({
      category: 'query',
      data: typeof url === 'string' ? { url } : undefined,
      level: 'info',
      timestamp: Date.now(),
    });

    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    // eslint-disable-next-line no-console
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    // eslint-disable-next-line no-console
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse as TParsedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = <T>(
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    } as unknown as Promise<UnknownData<T>>;
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    }as unknown as Promise<UnknownData<T>>;
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    }as unknown as Promise<UnknownData<T>>;
  }
};

const useData = <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeadersInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};

export default useData;
