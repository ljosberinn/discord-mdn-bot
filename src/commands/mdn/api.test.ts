import { updatedQueryBuilder } from './api';

import useData from '../../utils/useData';
import { getChosenResult } from '../../utils/discordTools';

const searchResponse = {
  query: 'document',
  locale: 'en-US',
  page: 1,
  pages: 383,
  start: 1,
  end: 10,
  next:
    'https://developer.mozilla.org/api/v1/search/en-US?highlight=false&page=2&q=document',
  previous: null,
  count: 3823,
  filters: [
    {
      name: 'Topics',
      slug: 'topic',
      options: [
        {
          name: 'APIs and DOM',
          slug: 'api',
          count: 2609,
          active: true,
          urls: {
            active: '/api/v1/search/en-US?highlight=false&q=document&topic=api',
            inactive: '/api/v1/search/en-US?highlight=false&q=document',
          },
        },
      ],
    },
  ],
  documents: [
    {
      title: 'Document directive',
      slug: 'Glossary/Document_directive',
      locale: 'en-US',
      excerpt:
        'CSP document directives are used in a Content-Security-Policy header and govern the properties of a document or worker environment to which a policy applies.',
    },
    {
      title: 'document environment',
      slug: 'Glossary/document_environment',
      locale: 'en-US',
      excerpt:
        "When the JavaScript global environment is a window or an iframe, it is called a document environment. A global environment is an environment that doesn't have an outer environment.",
    },
    {
      title: 'DOM (Document Object Model)',
      slug: 'Glossary/DOM',
      locale: 'en-US',
      excerpt:
        'The DOM (Document Object Model) is an API that represents and interacts with any HTML or XML document. The DOM is a document model loaded in the browser and representing the document as a node tree, where each node represents part of the document (e.g. an element, text string, or comment).',
    },
    {
      title: 'Archived open Web documentation',
      slug: 'Archive/Web',
      locale: 'en-US',
      excerpt:
        'The documentation listed below is archived, obsolete material about open Web topics.',
    },
    {
      title: 'Document.documentElement',
      slug: 'Web/API/Document/documentElement',
      locale: 'en-US',
      excerpt:
        'Document.documentElement returns the Element that is the root element of the document (for example, the html element for HTML documents).',
    },
    {
      title: 'Document.documentURI',
      slug: 'Web/API/Document/documentURI',
      locale: 'en-US',
      excerpt:
        'The documentURI read-only property of the Document interface returns the document location as a string.',
    },
    {
      title: 'Document.documentURIObject',
      slug: 'Web/API/Document/documentURIObject',
      locale: 'en-US',
      excerpt:
        'The Document.documentURIObject read-only property returns an nsIURI object representing the URI of the document.',
    },
    {
      title: 'Document',
      slug: 'Web/API/Document',
      locale: 'en-US',
      excerpt:
        "The Document interface represents any web page loaded in the browser and serves as an entry point into the web page's content, which is the DOM tree.",
    },
    {
      title: 'Document()',
      slug: 'Web/API/Document/Document',
      locale: 'en-US',
      excerpt:
        "The Document constructor creates a new Document object that is a web page loaded in the browser and serving as an entry point into the page's content.",
    },
    {
      title: '@document',
      slug: 'Web/CSS/@document',
      locale: 'en-US',
      excerpt:
        'The @document CSS at-rule restricts the style rules contained within it based on the URL of the document. It is designed primarily for user-defined style sheets, though it can be used on author-defined style sheets, too.',
    },
  ],
};

describe('updatedMDNQuery', () => {
  const mockUseData: jest.MockedFunction<typeof useData> = jest.fn();
  const mockChoose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  const editMsg = {
    edit: jest.fn(),
  };
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    reply: replyMock,
  };

  test('should work', async () => {
    mockUseData.mockResolvedValueOnce({
      error: false,
      text: null,
      json: searchResponse,
    });

    mockChoose.mockResolvedValueOnce({
      title: 'DOM (Document Object Model)',
      slug: 'Glossary/DOM',
      locale: 'en-US',
      excerpt:
        'The DOM (Document Object Model) is an API that represents and interacts with any HTML or XML document. The DOM is a document model loaded in the browser and representing the document as a node tree, where each node represents part of the document (e.g. an element, text string, or comment).',
    });

    sendMock.mockResolvedValue(editMsg);
    const handler = updatedQueryBuilder(mockUseData, mockChoose);

    await handler(msg, 'Search Term');
    expect(msg.channel.send.mock.calls).toMatchSnapshot();
    expect(editMsg.edit.mock.calls).toMatchSnapshot();
  });
});
