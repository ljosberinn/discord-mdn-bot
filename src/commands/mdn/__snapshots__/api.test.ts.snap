// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updatedMDNQuery should work 1`] = `
Array [
  Array [
    Object {
      "embed": Object {
        "author": null,
        "color": 8638706,
        "description": "1. [**Document directive** - CSP document directives are used in a Conten...](https://developer.mozilla.orgGlossary/Document_directive)
2. [**document environment** - When the JavaScript global environment is ...](https://developer.mozilla.orgGlossary/document_environment)
3. [**DOM (Document Object Model)** - The DOM (Document Object Model) is ...](https://developer.mozilla.orgGlossary/DOM)
4. [**Archived open Web documentation** - The documentation listed below ...](https://developer.mozilla.orgArchive/Web)
5. [**Document.documentElement** - Document.documentElement returns the E...](https://developer.mozilla.orgWeb/API/Document/documentElement)
6. [**Document.documentURI** - The documentURI read-only property of the ...](https://developer.mozilla.orgWeb/API/Document/documentURI)
7. [**Document.documentURIObject** - The Document.documentURIObject read-...](https://developer.mozilla.orgWeb/API/Document/documentURIObject)
8. [**Document** - The Document interface represents any web page loaded ...](https://developer.mozilla.orgWeb/API/Document)
9. [**Document()** - The Document constructor creates a new Document obje...](https://developer.mozilla.orgWeb/API/Document/Document)
10. [**@document** - The @document CSS at-rule restricts the style rules c...](https://developer.mozilla.orgWeb/CSS/@document)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`‚ùå\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
        "fields": Array [],
        "footer": Object {
          "iconURL": "https://avatars0.githubusercontent.com/u/7565578",
          "text": "10 results found",
        },
        "title": "MDN results for *Search Term*",
        "url": "https://developer.mozilla.org/en-US/search?q=Search%20Term",
      },
    },
  ],
]
`;

exports[`updatedMDNQuery should work 2`] = `
Array [
  Array [
    "https://developer.mozilla.org/Glossary/DOM",
    Object {
      "embed": null,
    },
  ],
]
`;
